test/common/router/router_fuzz_test.cc

```diff
@@ -0,0 +1,482 @@
+  void testDoNotForward(absl::optional<Http::LowerCaseString> not_forwarded_header_name);
+
+  Event::SimulatedTimeSystem test_time_;
+  std::string upstream_zone_{"to_az"};
+  envoy::config::core::v3::Locality upstream_locality_;
+  envoy::config::core::v3::HttpProtocolOptions common_http_protocol_options_;
+  NiceMock<Stats::MockIsolatedStatsStore> stats_store_;
+  NiceMock<Upstream::MockClusterManager> cm_;
+  NiceMock<Runtime::MockLoader> runtime_;
+  NiceMock<Random::MockRandomGenerator> random_;
+  Envoy::ConnectionPool::MockCancellable cancellable_;
+  Http::ContextImpl http_context_;
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> callbacks_;
+  MockShadowWriter* shadow_writer_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  FilterConfig config_;
+  RouterTestFilter router_;
+  Event::MockTimer* response_timeout_{};
+  Event::MockTimer* per_try_timeout_{};
+  Event::MockTimer* max_stream_duration_timer_{};
+  Network::Address::InstanceConstSharedPtr host_address_{
+      Network::Utility::resolveUrl("tcp://10.0.0.5:9211")};
+  NiceMock<Http::MockRequestEncoder> original_encoder_;
+  NiceMock<Http::MockRequestEncoder> second_encoder_;
+  NiceMock<Network::MockConnection> connection_;
+  Http::ResponseDecoder* response_decoder_ = nullptr;
+  Http::TestRequestHeaderMapImpl default_request_headers_;
+  Http::ResponseHeaderMapPtr redirect_headers_{
+      new Http::TestResponseHeaderMapImpl{{":status", "302"}, {"location", "http://www.foo.com"}}};
+  NiceMock<Tracing::MockSpan> span_;
+  NiceMock<StreamInfo::MockStreamInfo> upstream_stream_info_;
+};
+
+class RouterTest : public RouterTestBase {
+public:
+  RouterTest() : RouterTestBase(false, false, Protobuf::RepeatedPtrField<std::string>{}) {
+    EXPECT_CALL(callbacks_, activeSpan()).WillRepeatedly(ReturnRef(span_));
+  };
+
+  void replay(const test::common::router::RouterTestCase& input) {
+    ENVOY_LOG_MISC(info, "{}", input.DebugString());
+
+    for (const auto& action : input.actions()) {
+      ENVOY_LOG_MISC(info, "{}", action);
+
+      // every request gets a new encoder/decoder/stream?
+      // getting a segfault when onPoolReady is called on the second iteration here
+      NiceMock<Http::MockRequestEncoder> encoder1;
+      Http::ResponseDecoder* response_decoder = nullptr;
+      EXPECT_CALL(cm_.conn_pool_, newStream(_, _))
+          .WillOnce(
+              Invoke([&](Http::ResponseDecoder& decoder, Http::ConnectionPool::Callbacks& callbacks)
+                         -> Http::ConnectionPool::Cancellable* {
+                response_decoder = &decoder;
+                // still not too sure what these callbacks are for / do
+                callbacks.onPoolReady(encoder1, cm_.conn_pool_.host_, upstream_stream_info_);
+                return nullptr;
+              }));
+
+      // this creates the response timeout timer
+      expectResponseTimerCreate();
+
+      Http::TestRequestHeaderMapImpl headers{{"x-envoy-retry-on", "5xx"},
```

Comment:
fuzzing these headers is probably a good idea -- there are already definitions for headers in `test/fuzz/common.proto` that you can import

Reply:
Sure good idea.  I should probably keep some though right like the `x-envoy-retry-on` so that I can test retries?

Reply:
might be worth a fuzzing dictionary